### ISSUE TITLE
where's the code?

### ISSUE THREAD

**Issue by gubatron** [hyperledger/hyperledger#1]

**Comment by lsseckman** [comment]
Savage

**Comment by gvillalta99** [comment]
:+1:

**Comment by buhrmi** [comment]
I propose this empty POJO to be the genesis ledger: `{ }`

**Comment by gubatron** [comment]
a tad disappointing all the news coverage, the bombastic website, and then you come here and all there is... is a read me.

**Comment by ebuchman** [comment]
Ack

**Comment by GeorgLink** [comment]
It looks like the TSC has not finished evaluating initial code proposals.

> Since announcing the intent to form in December, the Hyperledger Project has received proposed code and technology contributions from several companies, including Blockstream, Digital Asset, IBM and Ripple. Other community members are contemplating contributions of their own.
> 
> As with all open source projects, technical contributions are welcome at any time, from anyone and will be reviewed by the newly formed Technical Steering Committee (TSC), which is comprised of industry-leading technical experts. The TSC is committed to open and transparent discussion, processes and decision-making. This group will oversee the project’s technical direction and working groups as well as manage multiple contributions to the code base. The TSC will be evaluating proposed contributions and go through an open community process to build out the initial and unified codebase.

https://www.hyperledger.org/news/announcement/2016/02/hyperledger-project-announces-30-founding-members

**Comment by akshithg** [comment]
Any idea about the language of the codebase?

**Comment by gubatron** [comment]
Thank you for clearing it up @GeorgLink

**Comment by jeffscottward** [comment]
I would like to see JavaScript.

Ethereum already has a VM implementation in JS, so you can interact with it
both browser-side or server-side.
Ethereum is also a permission-less ecosystem, that has an incentivization
mechanic.
I personally think it would be foolish not to follow that lead given it
will likely have the largest amount of eyes on reviewing the code.
IPFS is also working on a JS module for both client-server ubiquity, which
has close ties with the Ethereum community.

Javascript has the massive NPM ecosystem of course and global development
mindshare.
On top of that NPM is considering implementing NPM on-top of IPFS.
https://twitter.com/adam_baldwin/status/674397381643776000

TL:DR - just make a token-less version of EthereumJS so businesses don't
need to compete for compute cycles for everyday things. Generally a lot
less work given it would just be a tokenless fork.

**Comment by triconic** [comment]
I guess there are two questions. At risk of being too simplistic an answer ...

1. What language will hyperledger be written in? i.e. the core functionality.

If you take a look at the readme, the language will be dependent on which company/companies contribute code. I get the impression that hyperledger will be a combination of functionality from several contributors, perhaps with a core from one. Candidates for languages would include:

C++ => the Bitcoin core is written mostly in C++. Companies like Ripple are also hiring for C++
Java => Digital Asset Holdings (another name mentioned) appears to be writing their code in Java based on their open positions for devs. They also (interestingly) like Scala.

Bottom line, for the back-end, you're going to need high speed code. So, think C++. Maybe some Java, but I just don't see a cross-industry platform resting on a Java core. Maybe that's just me.

2. What language will you be able to write your app in that talks to the hyperledger?

You will likely see sample (reference) implementations in several other languages - as in all the typical app-y development languages (.NET, Java, Ruby, Python, Node ... pick a language). Most of the development around hyperledger will be at this level, not at the core level.

But, this is speculation. I'm looking forward to seeing some code also.

**Comment by ccerrato147** [comment]
I agree with @triconic C++ for the core itself and multiple clients to access it and build apps

**Comment by akshithg** [comment]
Go Lang would be a great candidate too. go-ethereum is a very mature and popular implementation. And IPFS too has a similar implementation. This could expose a JSON RPC. We can then have a client library in JS, like the web3 equivalent of ethereum, which can talk to the RPC from the browser or a server(node js). Client-server ubiquity as @jeffscottward stated.

**Comment by asanz** [comment]
+ethereum

**Comment by Subby83** [comment]
I would also "go" for a go implementation. :-)
Go is very strong when it comes to network or distributed applications since it can handle concurrency very well out of the box (goroutines, channels). It has a low footprint and is very performant.
As mentioned go-ethereum project is a great example.

**Comment by gubatron** [comment]
the problem with go is that it doesn't play very well with creating native bindings for other languages.
Also the tooling for go is non-existent compared to more mature languages.

**Comment by Subby83** [comment]
Okay, Scala then! :-)

**Comment by ameadors** [comment]
I second Scala!

On Feb 10, 2016, at 5:27 PM, Alexander Gehres <notifications@github.com<mailto:notifications@github.com>> wrote:


Okay, Scala then! :-)

**Comment by andrearota** [comment]
+ Java for me! :+1:

**Comment by joshuajeeson** [comment]
I believe Hyperledger from Asset holdings is currently in Erlang. Let's see
I guess.

**Comment by gubatron** [comment]
People who have done Scala with large code bases end up regreting it.
The time you save coding, you lose on super slow builds. I won't name which company had to invest on a 1000 server compile farm to deal with this issues. I will just say they are one of the most prominent Scala community voices and by looking at their stock price, if they go to hell so will Scala as they've already started pulling back from it.

My personal issue with Scala is the degree of freedom with creating your own DSLs leads up to having to almost learn a new language with every project you join. Everything almost works like magic, and if you let people have too much freedom you end up with this unmaintainable mess that in the long run takes forever to compile. If you want to go with a virtual machine, just go directly with Java, now with Java8 a lot of the good stuff from Scala is already available and Java9 should be coming up soon with more.

I'm fine with simple and solid foundations like C++ or Java. Then everything else can plug to these.
But in the case like something like Blockchain technology I'd have to favor C++ when the time comes to get really creative and optimize with things like custom memory allocation, advanced low level IO optimizations and things that you'd ultimately end up doing with C++/JNI/Java anyways. The only downside being that you have to build for every architecture out there and you wouldn't have to if you went 100% Java.

So I'd say C/C++.

**Comment by ccerrato147** [comment]
@gubatron +1

**Comment by jaycarey** [comment]
I'd chose [Kotlin](https://kotlinlang.org/):
 - First class tooling.
 - Comparable compile times to java.
 - Flawless interoperability with the entirety of the existing Java ecosystem.
 - Many of the best features of Scala (and others) but chosen by industry led pragmatism rather than academia.
 - Similarities with Java mean low the barriers to entry.

**Comment by gubatron** [comment]
@jaycarey thanks for sending this our way. Gives a good impression at first sight, loved your 4th bullet ;), that tends to be a real PITA in the Scala community, linguists gone wild.

**Comment by gubatron** [comment]
ha, and it's already good to be used with Jetbrains toolset.

**Comment by gubatron** [comment]
interesting, NPE checks at compile time too.

**Comment by jaycarey** [comment]
Another interesting feature that's not been pushed as much as the JVM side of things is that it can also compile down to javascript. This opens up some interesting doors with regards to portability and code reuse in a web app. I hope the industry catches on before long, it's my favourite language to use in my spare time!

**Comment by gazbert** [comment]
Java gets my vote, but learning a new language is always fun!

**Comment by morenoh149** [comment]
@joshuajeeson indeed I believe it's here https://github.com/danoprey/depreciated

**Comment by mileschet** [comment]
I would go for Java, but any language should be ok !

**Comment by ebuchman** [comment]
Funny that the base assumption in all this is that there is no code yet, so we can still pick a language. I believe the IBM version of the project has gone with Golang. Golang is an outstanding choice for this project, as so much new distributed systems infrastructure is being written in it (docker, etcd, consul, cloudflare, etc.). Also go's tooling is wonderful - those who think otherwise haven't used it ;)

**Comment by buhrmi** [comment]
yes but where is it? afaik the readme was going to be updated to point to
the current state of development (the ibm code base being one of the
proposals)

**Comment by mileschet** [comment]
@ebuchman go-lang is a great choice too ... +1

**Comment by buhrmi** [comment]
I think it should be javascript at the core because Atwood's Law states that `any application that can be written in JavaScript, will eventually be written in JavaScript.` http://blog.codinghorror.com/the-principle-of-least-power . and I agree with this statement. So I'd rather go with javascript and other people can plug into the network using their language-of-choice instead of using an other language at first and then somebody comes along with the inevitable hyperledger-js project

**Comment by jeffscottward** [comment]
Exactly what Stefan said. Hyperledger-js would eventually happen. Node.js
is super fast (comparable to go) and has a tremendous community. I do
believe there should be multiple clients in the way Ethereum is getting
built with a Haskell, Rust, Java, Go, C++, and JavaScript version but
ultimately we know which one over the very long haul is gonna win out, not
because it's the best language or runtime but because of access and
community. It's why Ethereum took off because it has an accessible language
(Solidity). I'm mostly a front end dev but I would be able to dig into
EthereumJS and understand it and how to use it.

**Comment by christo4ferris** [comment]
I think that as far as smart contracts are concerned, we should support whatever the developer chooses. Same for SDKs to interact with the various APIs of the components of the fabric.

However, for the runtime fabric itself, that's an entirely separate matter. Dealing with concurrency, number crunching and getting the most out of performance will be fundamentally important. I think Go offers up the best of those characteristics.  I certainly wouldn't advocate implementing in Javascript. Just because you can, doesn't mean you should.

**Comment by jeffscottward** [comment]
Just because you can is what has made JavaScript the #1 language in the
world. When there is a will there is a way and those issues will be solved
even if hoops are going to have to be jumped through. I would argue we
should because of the community and ubiquity.

**Comment by davidwowa** [comment]
I think the program language must be independent from platform, but fast and easy to implement for app-development and browser functionality, I propose JavaScript, Golang and Java/Scala.

**Comment by jtremback** [comment]
I think it should be a socket protocol so that anyone can use whatever language they want, like this: http://tendermint.com/posts/tendermint-socket-protocol/

**Comment by gubatron** [comment]
it looks like GO [has already been chosen](https://github.com/openblockchain/obc-peer)

**Comment by christo4ferris** [comment]
@gubatron to be clear, IBM chose Go for our implementation, as you have noted, but that doesn't mean that we (the Hyperledger technical community and TSC) have made the same choice. Certainly, I do think that would be the best choice, as we had suggested in the IBM proposal @binhn sent last week. However, the TSC has not yet made the decision.
